aiida_quantumespresso/calculations/pp.py

```diff
@@ -1,27 +1,218 @@
+        # Output datafile related exceptions
+        spec.exit_code(330, 'ERROR_OUTPUT_DATAFILE_MISSING',
+            message='The retrieved folder did not contain the required formatted data output file.')
+        spec.exit_code(331, 'ERROR_OUTPUT_DATAFILE_READ',
+            message='The formatted data output file could not be read.')
+        spec.exit_code(332, 'ERROR_UNSUPPORTED_DATAFILE_FORMAT',
+            message='The data file format is not supported by the parser')
+
+    def prepare_for_submission(self, folder): # pylint: disable=too-many-branches,too-many-statements
+        """Create the input files from the input nodes passed to this instance of the `CalcJob`.
+
+        :param folder: an `aiida.common.folders.Folder` to temporarily write files on disk
+        :return: `aiida.common.datastructures.CalcInfo` instance
+        """
+
+        # Put the first-level keys as uppercase (i.e., namelist and card names) and the second-level keys as lowercase
+        if 'parameters' in self.inputs:
+            parameters = _uppercase_dict(self.inputs.parameters.get_dict(), dict_name='parameters')
+            parameters = {k: _lowercase_dict(v, dict_name=k) for k, v in six.iteritems(parameters)}
+        else:
+            parameters = {}
+
+        # Set default values. NOTE: this is different from PW/CP
+        for blocked in self._blocked_keywords:
+            namelist = blocked[0].upper()
+            key = blocked[1].lower()
+            value = blocked[2]
+
+            if namelist in parameters:
+                if key in parameters[namelist]:
+                    raise exceptions.InputValidationError(
+                        "You cannot specify explicitly the '{}' key in the '{}' "
+                        'namelist.'.format(key, namelist))
+            else:
+                parameters[namelist] = {}
+            parameters[namelist][key] = value
+
+        # Check for essential keys
+        try:
+            plot_num = parameters['INPUTPP']['plot_num']
+        except KeyError:
+            raise exceptions.InputValidationError("'[INPUTPP][plot_num]' must be explicitly set")
+
+        try:
+            dimension = parameters['PLOT']['iflag']
+        except KeyError:
+            raise exceptions.InputValidationError("'[PLOT][iflag]' must be explicitly set")
+
+        # Check that a valid plot type is requested
+        if plot_num in range(23) and plot_num not in [14, 15, 16]: # Must be integer in range 0-22, but not 14-16:
+            parameters['INPUTPP']['plot_num'] = int(plot_num) # If this test passes, we can safely cast to int
+        else:
+            raise exceptions.InputValidationError("'plot_num' must be an integer in the range 0-23")
+
+        # Check for valid plot dimension:
+        if dimension in range(5): # Must be in range 0-4:
+            parameters['PLOT']['iflag'] = int(dimension)
+        else:
+            raise exceptions.InputValidationError("'iflag' must be an integer in the range 0-4")
```

Comment:
Might it be a good idea to move this to a validator for the `parameters` input? This way a user won't end up with a failed calculation in the database but the problem will already be caught at process input validation. Writing the validator is easy:
```
def validate_parameters(value, ctx):
    try:
        plot_num = value['INPUTPP']['plot_num']
    except KeyError:
        return "parameter `INPUTPP.plot_num` must be explicitly set"


spec.input('parameters', valid_type=orm.Dict, validator=validate_parameters)
```

Reply:
Done.